File: CPP/Lesson_234_reverse_nodes_in_k_group.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 234 -- Reverse Nodes In K Group
// Category   : Linked List
// Difficulty : Hard
// Study Plan : Day 117
// =============================================================
//
// QUESTION:
//   Reverse nodes in groups of k. Remaining tail stays.
// =============================================================
#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
struct N{int v;N* n;N(int v):v(v),n(nullptr){}};
N* reverseK(N* head,int k){N* d=new N(0);d->n=head;N* g=d;while(true){N* kth=g;for(int i=0;i<k;i++){kth=kth->n;if(!kth)return d->n;}N* nxt=kth->n;N* pre=nxt;N* cur=g->n;while(cur!=nxt){N* t=cur->n;cur->n=pre;pre=cur;cur=t;}N* tmp=g->n;g->n=kth;g=tmp;}}
N* to(vector<int> a){N* d=new N(0);N* c=d;for(int x:a){c->n=new N(x);c=c->n;}return d->n;}
int main(){auto r=reverseK(to({1,2,3,4,5}),2);while(r){cout<<r->v<<" ";r=r->n;}cout<<"\n";}

File: Python/Lesson_234_reverse_nodes_in_k_group.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 234 -- Reverse Nodes In K Group
# Category   : Linked List
# Difficulty : Hard
# Study Plan : Day 117
# =============================================================
#
# QUESTION:
#   Reverse nodes in groups of k. Remaining tail stays.
# =============================================================
class N:
    def __init__(s,v,n=None): s.v=v; s.n=n
def reverseK(head,k):
    d=N(0,head); g=d
    while True:
        kth=g
        for _ in range(k):
            kth=kth.n
            if not kth: return d.n
        nxt=kth.n; pre,cur=nxt,g.n
        while cur is not nxt:
            tmp=cur.n; cur.n=pre; pre=cur; cur=tmp
        tmp=g.n; g.n=kth; g=tmp
def to(a):
    d=N(0); c=d
    for x in a: c.n=N(x); c=c.n
    return d.n
def out(h):
    r=[]
    while h: r.append(h.v); h=h.n
    return r
if __name__=="__main__":
    print(out(reverseK(to([1,2,3,4,5]),2)))
